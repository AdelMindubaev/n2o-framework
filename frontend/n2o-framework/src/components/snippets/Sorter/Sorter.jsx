import React from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'

const SORT_TYPE = {
    ASC: 'ASC',
    DESC: 'DESC',
    NONE: 'NONE',
}

/**
 * Компонент сортировки. Оборачивает элемент, добавляя логику и иконки
 * @reactProps {element} children - Потомки компонента, которые будут обернуты
 * @reactProps {string} title - подска при наведение
 * @reactProps {string} sorting - текущее направление сортировки (ASC, DESC, NONE)
 * @reactProps {number} columnKey - идентификатор колонки
 * @reactProps {function} onSort - callback вызывается при смене сортировки
 * @example
 * <Sorter title="Hello"
 *             sorting="ASC"
 *             onSort={this.changeSortDirection}>
 *  World
 * </Sorter>
 */
class Sorter extends React.Component {
    constructor(props) {
        super(props)
        this.handleClick = this.handleClick.bind(this)
    }

    /**
   * Обработчик нажатия и вызов callback onSort
   * @param e
   */
    handleClick(e) {
        e.preventDefault()
        const { sorting, columnKey, onSort } = this.props
        let direction
        switch (sorting) {
            case SORT_TYPE.ASC:
                direction = SORT_TYPE.DESC
                break
            case SORT_TYPE.DESC:
                direction = SORT_TYPE.NONE
                break
            case SORT_TYPE.NONE:
                direction = SORT_TYPE.ASC
                break
            default:
                direction = SORT_TYPE.NONE
        }
        onSort(columnKey, direction)
    }

    /**
   * Базовый рендер компонента
   */
    render() {
        const { title, children, sorting } = this.props
        const iconClass = cx({
            'fa fa-sort-amount-asc': sorting === SORT_TYPE.ASC,
            'fa fa-sort-amount-desc': sorting === SORT_TYPE.DESC,
        })
        return (
            <a href="#" title={title} tabIndex={-1} onClick={this.handleClick}>
                {children}
                {' '}
                {iconClass && (
                    <i className={cx('n2o-sorting-icon', iconClass)} aria-hidden="true" />
                )}
            </a>
        )
    }
}

Sorter.propTypes = {
    children: PropTypes.node.isRequired,
    title: PropTypes.string,
    sorting: PropTypes.oneOf(Object.keys(SORT_TYPE)),
    columnKey: PropTypes.string,
    onSort: PropTypes.func,
}

Sorter.defaultProps = {
    sorting: 'NONE',
}

export default Sorter
