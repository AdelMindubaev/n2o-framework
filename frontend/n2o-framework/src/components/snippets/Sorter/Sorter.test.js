import React from 'react'

import Sorter from './Sorter'

describe('<Sorter />', () => {
    it('проверяет рендер по умолчанию', () => {
        const wrapper = shallow(<Sorter>Лейбл</Sorter>)
        expect(wrapper).toMatchSnapshot()
    })

    it('проверяет работы вызова callback при смене сортировки', () => {
        const onSort = jest.fn()
        const wrapper = mount(
            <Sorter sortDirection="NONE" columnKey="test" onSort={onSort}>
        Лейбл
            </Sorter>,
        )
        wrapper.simulate('click')
        expect(onSort).toHaveBeenCalledWith('test', 'ASC')
        wrapper.setProps({ sorting: 'ASC' })
        onSort.mockReset()
        wrapper.simulate('click')
        expect(onSort).toHaveBeenCalledWith('test', 'DESC')
        wrapper.setProps({ sorting: 'DESC' })
        onSort.mockReset()
        wrapper.simulate('click')
        expect(onSort).toHaveBeenCalledWith('test', 'NONE')
    })

    it('проверяет различные направление сортировки', () => {
        const wrapper = shallow(<Sorter sortDirection="NONE">Лейбл</Sorter>)
        expect(wrapper).toMatchSnapshot()
        wrapper.setProps({ sorting: 'ASC' })
        expect(wrapper).toMatchSnapshot()
        wrapper.setProps({ sorting: 'DESC' })
        expect(wrapper).toMatchSnapshot()
    })

    it('проверяет работу title', () => {
        const wrapper = shallow(<Sorter title="Супер">Лейбл</Sorter>)
        expect(wrapper.prop('title')).toEqual('Супер')
    })
})
