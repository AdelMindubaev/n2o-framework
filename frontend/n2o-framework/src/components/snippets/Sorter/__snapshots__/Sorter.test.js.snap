// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sorter /> проверяет различные направление сортировки 1`] = `
<a
  href="#"
  onClick={[Function]}
  tabIndex={-1}
>
  Лейбл
   
</a>
`;

exports[`<Sorter /> проверяет различные направление сортировки 2`] = `
<a
  href="#"
  onClick={[Function]}
  tabIndex={-1}
>
  Лейбл
   
  <i
    aria-hidden="true"
    className="n2o-sorting-icon fa fa-sort-amount-asc"
  />
</a>
`;

exports[`<Sorter /> проверяет различные направление сортировки 3`] = `
<a
  href="#"
  onClick={[Function]}
  tabIndex={-1}
>
  Лейбл
   
  <i
    aria-hidden="true"
    className="n2o-sorting-icon fa fa-sort-amount-desc"
  />
</a>
`;

exports[`<Sorter /> проверяет рендер по умолчанию 1`] = `
<a
  href="#"
  onClick={[Function]}
  tabIndex={-1}
>
  Лейбл
   
</a>
`;
